### PR TITLE
feat: Add --logformat switch to API server, repository server and controller (#3408)

### LINKED ISSUE #3408: Format control  plane component log output
# Summary

Similar as setting log level for each component, i want to be able to format the log output as json.

Current situation offer no possibility to change log output.

# Motivation

I want to monitor argo control plane containers and have a well structured log output so that i can easily integrate with elasticsearch.

# Proposal

As logging is implemented with logrus package which does support json format , need  implement in code for each component , the possibility to choose either text or json format for log output.

### PR DESCRIPTION
**Pull request by shuheiktgw**
Fixes https://github.com/argoproj/argo-cd/issues/3408

Enables to output log in JSON. Thanks for your review!

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [x] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)).

### CODE REVIEW

**Comment by jannfis** [comment]
Hi @shuheiktgw. many thanks for your contribution!

Unfortunately, your change doesn't pass the End-to-End tests:
```
time="2020-05-04T07:01:05Z" level=error msg="`../../dist/argocd account list --plaintext --server localhost:8080 --auth-token eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpYXQiOjE1ODg1NzU2NjUsImlzcyI6ImFyZ29jZCIsIm5iZiI6MTU4ODU3NTY2NSwic3ViIjoiYWRtaW4ifQ.LjIGH473x-j0_0mEGuIOx2iTicpnxbgRW4q2sAt5eG4 --insecure` failed exit status 1: time=\"2020-05-04T07:01:05Z\" level=fatal msg=\"not a valid logrus Level: \\\"text\\\"\"" execID=6PPSd
time="2020-05-04T07:01:05Z" level=fatal msg="`../../dist/argocd account list --plaintext --server localhost:8080 --auth-token eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpYXQiOjE1ODg1NzU2NjUsImlzcyI6ImFyZ29jZCIsIm5iZiI6MTU4ODU3NTY2NSwic3ViIjoiYWRtaW4ifQ.LjIGH473x-j0_0mEGuIOx2iTicpnxbgRW4q2sAt5eG4 --insecure` failed exit status 1: time=\"2020-05-04T07:01:05Z\" level=fatal msg=\"not a valid logrus Level: \\\"text\\\"\""
```

That CLI call doesn't explicitly set log level or format, so there might be a logic error somewhere in your code.

**Comment by shuheiktgw** [comment]
Thank you for your comment, and sorry for not noticing the failing e2e tests! Let me check what is wrong with this code! 👍

File: util/cli/cli.go [hunk 1]
```diff
@@ -137,14 +137,23 @@ func ReadAndConfirmPassword() (string, error) {
 	}
 }
 
+// SetLogFormat sets a logrus log format
+func SetLogFormat(logFormat string) {
```
Comment (jannfis) [inline util/cli/cli.go:141]: I think this method should explicitly test for valid log formats. I.e. in the flag, it says:

> Set the logging format. One of: text|json

But actually, you can give it any value.

**Review by jannfis — CHANGES_REQUESTED** [review]
Thanks a lot for fixing it, @shuheiktgw. It looks good to me so far, I just found one thing that needs to be changed IMO.

Also, we can ignore the unauthorized SonarQube run. That's something we still have to investigate, i.e. why it authorizes sometimes and sometimes not.

File: util/cli/cli.go (hunk 1 above)
Comment (shuheiktgw) [reply util/cli/cli.go:141]: Thanks for your comment! I fixed to output log when the given log format is invalid (just like `errors.CheckError` does in `SetLogLevel` method)!

**Review by jannfis — APPROVED** [review]
LGTM! Thank you for this contribution!

**Comment by AndreiBamboi** [comment]
Thanks for your effort.

**Pull request merged**